[local-preview] Add Windows and Mac support (#2349)
gitpod-io/website: gitpod/docs/self-hosted/latest/local-preview.md

This install method runs a `k3s` cluster inside a Docker container. Self-signed certificates are automatically created and a Gitpod instance using these certificates will be installed into the `k3s` cluster. See our [TLS](./advanced/tls) page for more information about Gitpod and certificates.

## 1. Running the docker container

First, Create a volume to which the Gitpod container will attach:

```bash
docker volume create gitpod
```

Run the following command to get the `local-preview` Docker container up and running:

```bash
docker run -p 443:443 --privileged --name gitpod --rm -it -v gitpod:/var/gitpod eu.gcr.io/gitpod-core-dev/build/local-preview
```

Unpacking the above command:

- `-p 443:443` to map the `443` container port to host.
- `--privileged` to be able to run docker (and hence `k3s`) inside the container. This is necessary.
- `--name gitpod` to set the name of the docker container for further access.
- `gitpod:/var/gitpod` to store the workspace files, etc into the `gitpod` volume.

> > **Note:** By default, `127-0-0-1.nip.io` is the DOMAIN to access Gitpod. To use another host network IP Address, the `DOMAIN` environment can be set accordingly via an `-e` flag in the above command. For Example, `192.168.0.42` Host Network IP would be `192-168-0-42.nip.io`, set by appending `-e DOMAIN=192-168-0-42.nip.io`. This is useful to share access to the Gitpod Self-Hosted instance running on your machine within your local network.

Reviewer: ```suggestion
> **Note:** By default, `127-0-0-1.nip.io` is the DOMAIN to access Gitpod. To use another host network IP Address, the `DOMAIN` environment can be set accordingly via an `-e` flag in the above command. This is useful to share access to the Gitpod Self-Hosted instance running on your machine within your local network. For Example, `192.168.0.42` Host Network IP would be `192-168-0-42.nip.io`, set by appending `-e DOMAIN=192-168-0-42.nip.io`. 
```
Switched ordering of the sentences because I think it is important to mention why this is useful earlier

# Local Preview of Gitpod Self-Hosted

> **Note:** Currently, this Local Preview of Gitpod Self-Hosted is in [beta](../../references/gitpod-releases). It does not work on M1 Mac. See [the relevant issue](https://github.com/gitpod-io/gitpod/issues/9075) for more information.

The Local Preview of Gitpod Self-Hosted is the easiest way to try out Gitpod locally without having to spin up a Kubernetes cluster - allowing you to try Gitpod Self-Hosted locally with minimal effort and resource requirements. As a result, this is **not intended for production** usage. Please refer to the [Getting Started](./getting-started) page for instructions on how to install Gitpod for continuous usage.

## Running the docker container

First, Create a volume to which the Gitpod container will attach.

```bash

> docker volume create gitpod

Reviewer: It seems like this is adding another step to the process. I'd love to understand why and if there is a way to circumvent it.

Author: Yep, This was added for WIndows. With Windows, As docker is ran inside a VM and directly through the Host. There's some path changes that we have to do specifically for Windows, and have separate section for that.

Instead we could rely on docker creating the volume, and using that across all platforms which we can be sure will work and also have a consistent experience across

Me and @corneliusludmann discussed on this, and thought if we could remove access to workspace data fully. this whole volume creation thing can be fully removed then as there isn't a lot of usefulness in storing workspace data separately on the host for user to access.

```

> This install method runs a `k3s` cluster inside a Docker container. Self-signed certificates are automatically created and a Gitpod instance using these certificates will be installed into the `k3s` cluster. See our [TLS](./advanced/tls) page for more information about Gitpod and certificates.

Reviewer: I don't follow this section. At first I was going to ask if this is a section that is just background on what you are actually installing. If this is the case we should move it to it's own section in order to not complicate the installation steps. However - Are we asking the user to get their own IP address here? Because at first I did not realise this.

As Gitpod needs this information in a domain format (and not just a IP address). For this to work, a `nip` URL can be created by replacing the `.` with `-`, and suffixing it with `nip.io`. For example, a private IP address like `10.0.0.7` becomes `10-0-0-7.nip.io`. This is then passed to the container through the `DOMAIN` environment variable.

Run the following command to get the `preview-install` Docker container up and running:

```bash
docker run -p 443:443 -e DOMAIN=10-0-0-7.nip.io --privileged --name gitpod --rm -it -v gitpod:/var/gitpod eu.gcr.io/gitpod-core-dev/build/preview-install
```

Unpacking the above command:

- `-p 443:443` to map the `443` container port to host.
- `-e DOMAIN=10-0-0-7.nip.io` to pass the DOMAIN to the Gitpod installation. [nip.io](https://nip.io/) is just wildcard DNS for local addresses, so all off this is local, and cannot be accessed over the internet.
- `--privileged` to be able to run docker (and hence `k3s`) inside the container. This is necessary.
- `--name gitpod` to set the name of the docker container for further access.
- `gitpod:/var/gitpod` to store the workspace files, etc into the `gitpod` volume.

## Accessing Gitpod

> As this is a self-signed instance of Gitpod, the gitpod root CA cert has to be imported into your browser manually to access the full functionality of Gitpod. The certificate can be retrieved by running

Author: > It's not just about "access the full functionality of Gitpod" right - you cannot access Gitpod at all

You can access Gitpod without improting the cert, by clicking I accept and Go forward warning in your browser, and you can complete the Oauth setting with that. It's the workspace access where you get a blank screen, with err's in your Browsers network dashboard.

My reason for specifying this is that, Some developers could think "I don't have to import the cert, I can just move forward on the warning" and it does work until the workspace access. Hence, the clearer mention there. 🤔

> As this is a self-signed instance of Gitpod, the gitpod root CA cert has to be imported into your browser manually to access the full functionality of Gitpod. The certificate can be retrieved by running

Reviewer: I understand - that makes sense let's keep it like this then!

> # Local Preview of Gitpod Self-Hosted

Reviewer: ```suggestion
# How to Install the Local Preview Gitpod Self-Hosted 
```
This is just for completeness and to indicate that this is an how to guide.

> > **Note:** Currently, this Local Preview of Gitpod Self-Hosted is in [beta](../../references/gitpod-releases). It does not work on M1 Mac. See [the relevant issue](https://github.com/gitpod-io/gitpod/issues/9075) for more information.

Reviewer: ```suggestion
> **Note:** Currently, this Local Preview of Gitpod Self-Hosted is in [beta](../../references/gitpod-releases). **It does not work on Macs with Apple Silicon (e.g. M1).** See [the relevant issue](https://github.com/gitpod-io/gitpod/issues/9075) for more information.
```

Author: Nice, Forgot that there was M2 coming in

The Local Preview of Gitpod Self-Hosted is the easiest way to try out Gitpod locally without having to spin up a Kubernetes cluster - allowing you to try Gitpod Self-Hosted locally with minimal effort and resource requirements. As a result, this is **not intended for production** usage. Please refer to the [Getting Started](./getting-started) page for instructions on how to install Gitpod for continuous usage.

> ## Running the docker container

Reviewer: ```suggestion
## 1. Running the docker container
```

> First, Create a volume to which the Gitpod container will attach.

Reviewer: ```suggestion
First, Create a volume to which the Gitpod container will attach:
```

```bash
docker volume create gitpod
```

> This install method runs a `k3s` cluster inside a Docker container. Self-signed certificates are automatically created and a Gitpod instance using these certificates will be installed into the `k3s` cluster. See our [TLS](./advanced/tls) page for more information about Gitpod and certificates.

Reviewer: This does not contain any steps one needs to execute right? As such, it is background information and I would move this paragraph to the second position - i.e. below the first paragraph above ("The Local Preview of Gitpod Self-Hosted is the easiest way [...]"

Run the following command to get the `local-preview` Docker container up and running:

```bash
docker run -p 443:443 --privileged --name gitpod --rm -it -v gitpod:/var/gitpod eu.gcr.io/gitpod-core-dev/build/local-preview
```

Unpacking the above command:

- `-p 443:443` to map the `443` container port to host.
- `--privileged` to be able to run docker (and hence `k3s`) inside the container. This is necessary.
- `--name gitpod` to set the name of the docker container for further access.
- `gitpod:/var/gitpod` to store the workspace files, etc into the `gitpod` volume.

> > **Note:** By default, `127-0-0-1.nip.io` is the DOMAIN to access Gitpod. To use an other host network IP Address, `DOMAIN` environment can be set accordingly through `-e` flag in the above command. For Example, `192.168.0.42` Host Network IP would be `192-168-0-42.nip.io`. This is useful to share the Gitpod access in your local network.

Reviewer: ```suggestion
> **Note:** By default, `127-0-0-1.nip.io` is the DOMAIN to access Gitpod. To use another host network IP Address, the `DOMAIN` environment can be set accordingly via an `-e` flag in the above command. For Example, `192.168.0.42` Host Network IP would be `192-168-0-42.nip.io`, set by appending `-e 192.168.0.42`. This is useful to share access to the Gitpod Self-Hosted instance running on your machine within your local network. 
```

1. I added an example usage of the `-e` flag - is that correct? 
2. Have we tested accessing the local preview via the local network from another machine?

Author: 1. It should be `-e DOMAIN=192-168-0-42.nip.io`, and I updated it accordingly
2. Not yet, but it should work as that behaviour is expected from Docker and how host networks work.

We can expand on 2, after we test and even try to make it simpler i.e using just the IP address, etc? This is aimed for advanced users.

> ## Accessing Gitpod

Reviewer: ```suggestion
## 2. Accessing Gitpod
```

> As this is a self-signed instance of Gitpod, the gitpod root CA cert has to be imported into your browser manually to access the full functionality of Gitpod. The certificate can be retrieved by running

Reviewer: ```suggestion
As this is a self-signed instance of Gitpod, the Gitpod root CA cert has to be imported into your browser manually to access the full functionality of Gitpod. The certificate can be retrieved by running the following:
```

```bash
docker cp gitpod:/var/gitpod/gitpod-ca.crt $HOME/gitpod-ca.crt
```

> This certificate at `$HOME/gitpod-ca.crt` can then be loaded into your browser. Most browsers also require a restart before they can start to use the imported certificate.

Reviewer: ```suggestion
This certificate is saved at `$HOME/gitpod-ca.crt` and can then be loaded into your browser. Most browsers also require a restart before they can start to use the imported certificate.
```

> Once the certificate is loaded, The URL to access the Gitpod instance would be `https://127-0-0-1.nip.io` unless the `DOMAIN` environment is overriden in which

Reviewer: ```suggestion
Once the certificate is loaded, the URL to access the Gitpod instance would be `https://127-0-0-1.nip.io` unless the `DOMAIN` environment is overridden in which
```

- `--name gitpod` to set the name of the docker container for further access

> - `/tmp/gitpod:/var/gitpod` to store the workspace files, etc into `tmp/gitpod` temporary directory

Author: need to change this.
